src/fam.rs

```diff
@@ -430,10 +430,31 @@ impl<T: Default + FamStruct> PartialEq for FamStructWrapper<T> {
 
 impl<T: Default + FamStruct> Clone for FamStructWrapper<T> {
     fn clone(&self) -> Self {
-        // It is safe to unwrap() for now since `self` is a valid `FamStructWrapper`, so the number
-        // of entries can't be > T::max_len(). In the next commit, unwrap() will not be needed
-        // anymore.
-        FamStructWrapper::from_entries(self.as_slice()).unwrap()
+        // The number of entries (self.as_slice().len()) can't be > T::max_len() since `self` is a
+        // valid `FamStructWrapper`.
+        let required_mem_allocator_capacity =
+            FamStructWrapper::<T>::mem_allocator_len(self.as_slice().len());
+
+        let mut mem_allocator = Vec::with_capacity(required_mem_allocator_capacity);
+
+        // This is safe as long as the requirements for the `FamStruct` trait to be safe are met
+        // (the implementing type and the entries elements are POD, therefore `Copy`, so memory
+        // safety can't be violated by the ownership of `fam_struct`). It is also safe because we're
+        // trying to read a T from a `&T` that is pointing to a properly initialized and aligned T.
+        unsafe {
+            let fam_struct: T = std::ptr::read(self.as_fam_struct_ref());
```

Comment:
what if the FAM header contains a pointer ? Can it happen ?

Reply:
Good question! As discussed offline, this might be possible, but `T` must be `Default` and from what I understand, implementing `Default` for a pointer becomes really awkward. Also, if a pointer is `Copy`, I wonder if `ptr::read` is actually unsafe :-?.
@jiangliu @alexandruag do you have some input on this? Thanks!

Reply:
`FamStruct` is an unsafe trait, which requires following conditions for safety
```
/// Trait for accessing properties of C defined FAM structures.
///
/// This is unsafe due to the number of constraints that aren't checked:
/// * the implementer should be a POD
/// * the implementor should contain a flexible array member of elements of type `Entry`
/// * `Entry` should be a POD
```
So it's illegal for an FamStruct implementor to host pointer fields.

Reply:
I was under the same impression, but after some googling/asking around, looks like a pointer is a POD. So isn't this actually right? :-?

Reply:
I like this definition from https://stackoverflow.com/questions/45634083/is-there-a-concept-of-pod-types-in-rust:
POD: Types whose values can be duplicated simply by copying bits.

Reply:
Sorry, my bad. Now I remember that the entire class was written with the assumption that `T` is memcopyable. It's just very strange that in C++ the following code prints true:
```
struct Pod {
    int *a;
};

int main()
{
    cout<<is_pod<Pod>::value<<endl;

    return 0;
}
```

Reply:
Hi! Don't know what authoritative source is best to quote here, but POD structs can contain pointers, which are effectively just integers meant to be interpreted as addresses. A pointer holding a value does not come with any guarantees by itself, and all safety concerns are passed on to whomever actually uses that pointer for a memory operation later (since all fields of a POD struct are always accessible and can be modified, any method operating on them must take into account that every possible value can be encountered). So, even if the POD struct contains pointers (which is unlikely), that doesn't seem to break any of the current assumptions.

Reply:
I think the place to go would be the [C++ standard](https://github.com/cplusplus/draft). I don't own the hard copy, but the latest snapshot from the official repo loosely defines POD types through the template @serban300 used -

```
is_pod<T> is a Cpp17UnaryTypeTrait (20.15.2) with a base characteristic of true_type if T is a POD
type, and false_type otherwise. A POD class is a class that is both a trivial class and a standard-layout
class, and has no non-static data members of type non-POD class (or array thereof). A POD type is a
scalar type, a POD class, an array of such a type, or a cv-qualified version of one of these types.

[Note 1 : It is unspecified whether a closure type (7.5.5.2) is a POD type. — end note]
```

So my understanding here is in line with @alexandruag in the sense that scalar types are inherently POD, and since a pointer is basically just a number, it falls into this category.

I also found an unofficial definition that (more helpfully IMO) defines POD types as anything without a vtable.

Reply:
Thanks for the clarifications. I agree, I don't have any concern here. I will resolve this comment.

Reply:
As a rule of thumb anything in C++ is POD that can be used from C. So any C struct is almost by definition POD.

Personally, I find [cppreference](https://en.cppreference.com/w/cpp/named_req/PODType) easier to read than the standard, but it's not authoritative.